pkg/controller/elasticsearch/services/services.go

```diff
@@ -181,26 +183,59 @@ func getServiceByName(c k8s.Client, es esv1.Elasticsearch, serviceName string) (
 	return svc, nil
 }
 
-// ElasticsearchURL calculates the base url for Elasticsearch, taking into account the currently running pods.
-// If there is an HTTP scheme mismatch between spec and pods we switch to requesting individual pods directly
-// otherwise this delegates to ExternalServiceURL.
-func ElasticsearchURL(es esv1.Elasticsearch, pods []corev1.Pod) string {
-	var schemeChange bool
-	for _, p := range pods {
-		scheme, exists := p.Labels[label.HTTPSchemeLabelName]
-		if exists && scheme != es.Spec.HTTP.Protocol() {
-			// scheme in existing pods does not match scheme in spec, user toggled HTTP(S)
-			schemeChange = true
+type urlProvider struct {
+	pods   func() []corev1.Pod
+	svcURL string
+}
+
+// PodURL implements client.URLProvider.
+func (u *urlProvider) PodURL() string {
+	var ready, running []corev1.Pod
+	for _, p := range u.pods() {
+		if k8s.IsPodReady(p) {
+			ready = append(ready, p)
 		}
-	}
-	if schemeChange {
-		// switch to sending requests directly to a random pod instead of going through the service
-		randomPod := pods[rand.Intn(len(pods))] //nolint:gosec
-		if podURL := ElasticsearchPodURL(randomPod); podURL != "" {
-			return podURL
+		if k8s.IsPodRunning(p) {
+			running = append(running, p)
 		}
 	}
-	return InternalServiceURL(es)
+	switch {
+	case len(ready) > 0:
+		return randomESPodURL(ready)
+	case len(running) > 0:
+		return randomESPodURL(running)
+	default:
+		return u.ServiceURL()
+	}
+}
+
+// ServiceURL implements client.URLProvider.
+func (u *urlProvider) ServiceURL() string {
+	return u.svcURL
+}
+
+// HasEndpoints implements client.URLProvider.
+func (u *urlProvider) HasEndpoints() bool {
+	return len(k8s.RunningPods(u.pods())) > 0
+}
+
+func NewElasticsearchURLProvider(ctx context.Context, es esv1.Elasticsearch, client k8s.Client) client.URLProvider {
+	return &urlProvider{
+		pods: func() []corev1.Pod {
+			log := ulog.FromContext(ctx)
+			pods, err := k8s.PodsMatchingLabels(client, es.Namespace, label.NewLabelSelectorForElasticsearch(es))
+			if err != nil {
+				log.Error(err, "while fetching pods from cache in URL provider")
```

Comment:
@barkbay with https://github.com/elastic/cloud-on-k8s/pull/7847/files/1649df75c87bdd0033acd79e2530661fc3c48e7e#r1623879698 you mean return here in case of an error?

Reply:
Yes